syntax = "proto3";

package meetmanager;

service MeetManagerService {
  // Meet Operations
  rpc GetMeets (Empty) returns (MeetList);
  rpc GetDashboardStats (Empty) returns (DashboardStats);

  // Team Operations
  rpc GetTeams (Empty) returns (TeamList);
  rpc GetTeam (TeamRequest) returns (Team);

  // Athlete Operations
  rpc GetAthletes (Empty) returns (AthleteList);
  rpc GetAthlete (AthleteRequest) returns (Athlete);
  
  // Event Operations
  rpc GetEvents (Empty) returns (EventList);

  // Admin Operations
  rpc ListDatasets (Empty) returns (DatasetList);
  rpc SetActiveDataset (DatasetRequest) returns (Empty);
  rpc UploadDataset (stream UploadRequest) returns (UploadResponse);

  // New Data Operations
  rpc GetRelays (Empty) returns (RelayList);
  rpc GetScores (Empty) returns (ScoreList);
  rpc GetEntries (Empty) returns (EntryList);
  
  // Session Operations
  rpc GetSessions (Empty) returns (SessionList);

  // Configuration
  rpc GetAdminConfig (Empty) returns (AdminConfig);
  rpc UpdateAdminConfig (AdminConfig) returns (AdminConfig);
  rpc GetEventScores (Empty) returns (EventScoreList);
  rpc GenerateReport (GenerateReportRequest) returns (ReportResponse);
}

message Dataset {
  string filename = 1;
  bool is_active = 2;
  string last_modified = 3;
}

message DatasetList {
  repeated Dataset datasets = 1;
}

message DatasetRequest {
  string filename = 1;
}

message UploadRequest {
  oneof data {
    string filename = 1;
    bytes chunk = 2;
  }
}

message UploadResponse {
  bool success = 1;
  string message = 2;
}

// Data Entities

message Relay {
  int32 id = 1; // Uniq combination of RELAY_no? Or just index?
  int32 event_id = 2; // Event_ptr
  int32 team_id = 3; // Team_no
  string team_name = 4;
  string leg1_name = 5;
  string leg2_name = 6;
  string leg3_name = 7;
  string leg4_name = 8;
  string seed_time = 9;
  string final_time = 10;
  int32 place = 11;
  string event_name = 12;
  string relay_letter = 13;
  int32 heat = 14;
  int32 lane = 15;
}

message RelayList {
  repeated Relay relays = 1;
}

message Score {
  int32 team_id = 1;
  string team_name = 2;
  float individual_points = 3;
  float relay_points = 4;
  float total_points = 5;
  int32 rank = 6;
  string meet_name = 7;
}

message ScoreList {
  repeated Score scores = 1;
}

message EventScore {
  int32 event_id = 1;
  string event_name = 2;
  repeated Entry entries = 3; // Reusing Entry which has place/points
}

message EventScoreList {
  repeated EventScore event_scores = 1;
}

message Entry {
  int32 id = 1; // Entry_no
  int32 event_id = 2; // Event_ptr
  int32 athlete_id = 3; // Ath_no
  string athlete_name = 4;
  int32 team_id = 5; // Team_no
  string team_name = 6;
  string seed_time = 7;
  string final_time = 8;
  int32 place = 9;
  string event_name = 10;
  int32 heat = 11;
  int32 lane = 12;
  float points = 14;
}

message EntryList {
  repeated Entry entries = 1;
}

message Session {
  string id = 1;
  string meet_id = 2;
  string name = 3;
  string date = 4;
  string warm_up_time = 5;
  string start_time = 6;
  int32 event_count = 7;
  int32 session_num = 8;
  int32 day = 9;
}

message SessionList {
  repeated Session sessions = 1;
}

message AdminConfig {
  string meet_name = 1;
  string meet_description = 2; // Details like location, dates
}

message Empty {}

message Meet {
  string id = 1;
  string name = 2;
  string location = 3;
  string start_date = 4;
  string end_date = 5;
  string status = 6; // active, upcoming, completed
}

message MeetList {
  repeated Meet meets = 1;
}

message DashboardStats {
  int32 meet_count = 1;
  int32 team_count = 2;
  int32 athlete_count = 3;
  int32 event_count = 4;
}

message Team {
  int32 id = 1;
  string name = 2;
  string code = 3; // Abbreviation
  string lsc = 4;
  string city = 5;
  string state = 6;
  int32 athlete_count = 7;
}

message TeamList {
  repeated Team teams = 1;
}

message TeamRequest {
  int32 id = 1;
}

message Athlete {
  int32 id = 1;
  string first_name = 2;
  string last_name = 3;
  string gender = 4; // F or M
  int32 age = 5;
  int32 team_id = 6;
  string team_name = 7;
  string school_year = 8;
  string reg_no = 9;
  string date_of_birth = 10;
}

message AthleteList {
  repeated Athlete athletes = 1;
}

message AthleteRequest {
  int32 id = 1;
}

message Event {
  int32 id = 1; // Event_no
  string gender = 2;
  int32 distance = 3;
  string stroke = 4;
  int32 low_age = 5;
  int32 high_age = 6;
  int32 session = 7;
  string status = 8;
  int32 entry_count = 9;
}

message EventList {
  repeated Event events = 1;
}

enum ReportType {
  PSYCH = 0;
  ENTRIES = 1;
  LINEUPS = 2;
  RESULTS = 3;
}

message GenerateReportRequest {
  ReportType type = 1;
  string title = 2;
  string team_filter = 3;
}

message ReportResponse {
  bool success = 1;
  string message = 2;
  bytes pdf_content = 3;
  string filename = 4;
}
